keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,935 @@
+	// ephemeral volumes are allocated.
+	EphemeralStorageClassName = "<ephemeral>"
+)
+```
+
+This approach has the advantage that the size of the `CSIStoragePool`
+objects do not depend on the number of nodes in the cluster for local
+storage.
+
+The downsides are:
+- Some attributes (driver name, topology) must be stored multiple times
+  compared to a single, more complex object that contains all information
+  for one CSI driver, so overall size in etcd is higher.
+- Higher number of objects which all need to be retrieved by a client
+  which does not already know which `CSIStoragePool` is is interested in.
+
+##### Example: local storage
+
+```
+apiVersion: storage.k8s.io/v1alpha1
+kind: CSIStoragePool
+metadata:
+  name: sp-ab96d356-0d31-11ea-ade1-8b7e883d1af1
+spec:
+  driverName: hostpath.csi.k8s.io
+status:
+  classes:
+  - capacity: 256G
+    storageClassName: <fallback>
+  nodes:
+  - node-1
+
+apiVersion: storage.k8s.io/v1alpha1
+kind: CSIStoragePool
+metadata:
+  name: sp-c3723f32-0d32-11ea-a14f-fbaf155dff50
+spec:
+  driverName: hostpath.csi.k8s.io
+status:
+  classes:
+  - capacity: 512G
+    storageClassName: <fallback>
+  nodes:
+  - node-2
+```
+
+##### Example: affect of storage classes
+
+```
+apiVersion: storage.k8s.io/v1alpha1
+kind: CSIStoragePool
+metadata:
+  name: sp-9c17f6fc-6ada-488f-9d44-c5d63ecdf7a9
+spec:
+  driverName: lvm
+status:
+  classes:
+  - capacity: 256G
+    storageClassName: striped
+  - capacity: 128G
+    storageClassName: mirrored
+  nodes:
+  - node-1
+```
+
+##### Example: network attached storage
+
+```
+apiVersion: storage.k8s.io/v1alpha1
+kind: CSIStoragePool
+metadata:
+  name: sp-b0963bb5-37cf-415d-9fb1-667499172320
+spec:
+  driverName: pd.csi.storage.gke.io
+status:
+  classes:
+  - capacity: 128G
+    storageClassName: <fallback>
+  nodeTopology:
+    nodeSelectorTerms:
+    - matchExpressions:
+      - key: failure-domain.beta.kubernetes.io/region
+        operator: In
+        values:
+        - us-east-1
+
+apiVersion: storage.k8s.io/v1alpha1
+kind: CSIStoragePool
+metadata:
+  name: sp-64103396-0d32-11ea-945c-e3ede5f0f3ae
+spec:
+  driverName: pd.csi.storage.gke.io
+status:
+  classes:
+  - capacity: 256G
+    storageClassName: <fallback>
+  nodeTopology:
+    nodeSelectorTerms:
+    - matchExpressions:
+      - key: failure-domain.beta.kubernetes.io/region
+        operator: In
+        values:
+        - us-west-1
+```
+
+#### CSIDriver.spec.storageCapacity
+
+A new field `storageCapacity` of type `boolean` with default `false`
+in
+[CSIDriver.spec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#csidriverspec-v1beta1-storage-k8s-io)
+indicates whether a driver deployment will create `CSIStoragePool`
+objects. This allows the scheduler to determine whether it needs to
+consider available capacity for volumes of that driver or can proceed
+without it (as it does now).
+
+
+### Updating capacity information with external-provisioner
```

Comment:
How often are we going to poll the driver, and how often are we going to update K8s objects? How do we ensure that a 5000 node cluster is not going to get overwhelmed with API updates?

Reply:
As I mentioned in the other comment, this is going to be tricky to get right, on the one hand you don't want to overwhelm the API server, on the other hand we need to make sure the scheduler is operating on fresh capacity info. I think caching the capacity in the scheduler may be necessary: basically the scheduler maintains and updates the capacity a local copy of CSIStoragePool that gets reconciled every once in a while with the copy in the api server